lib/vagrant/util/downloader.rb

```diff
@@ -178,6 +191,47 @@ def head
 
       protected
 
+      # Apply any checksum validations based on provided
+      # options content
+      #
+      # @param source [String] Source of file
+      # @param path [String, Pathname] local file path
+      # @param checksums [Hash] User provided options
+      # @option checksums [String] :md5 Compare MD5 checksum
+      # @option checksums [String] :sha1 Compare SHA1 checksum
+      # @return [Boolean]
+      def validate_download!(source, path, checksums)
+        CHECKSUM_MAP.each do |type, klass|
+          if checksums[type]
+            result = checksum_file(klass, path)
+            if checksums[type] != result
+              raise Errors::DownloaderChecksumError.new(
+                source: source,
+                path: path,
+                expected_checksum: checksums[type],
+                actual_checksum: result
+              )
+            end
+          end
+        end
+        true
+      end
+
+      # Generate checksum on given file
+      #
+      # @param digest_class [Class] Digest class to use for generating checksum
+      # @param path [String, Pathname] Path to file
+      # @return [String] hexdigest result
+      def checksum_file(digest_class, path)
+        digester = digest_class.new
+        File.open(path.to_s, 'rb') do |file|
```

Comment:
This is 1024 _bytes_, which seems really small. The [core method](https://github.com/ruby/ruby/blob/9ec0f8b7cb49c538f2bed77ec1506d59119f4bf1/ext/digest/lib/digest.rb#L49-L57) uses 16384...

Also, there is a core method for this 😄